### PR TITLE
Handle with cookies on flight-sql-client in the JDBC driver

### PR DESCRIPTION
**Pull request by jcralmeida**

### CODE REVIEW

File: java/flight/flight-jdbc-driver/src/main/java/org/apache/arrow/driver/jdbc/client/ArrowFlightSqlClientHandler.java [hunk 1]
```diff
@@ -515,7 +516,8 @@ public ArrowFlightSqlClientHandler build() throws SQLException {
           clientBuilder.trustedCertificates(
               ClientAuthenticationUtils.getCertificateStream(keyStorePath, keyStorePassword));
         }
-        final FlightClient client = clientBuilder.build();
+        ClientCookieMiddleware.Factory factory = new ClientCookieMiddleware.Factory();
```
Comment (jduo) [inline java/flight/flight-jdbc-driver/src/main/java/org/apache/arrow/driver/jdbc/client/ArrowFlightSqlClientHandler.java:519]: Nit: Should use withMiddlewareFactories and do so before line 506 to be inconsistent with how we're adding middleware factories elsewhere.

File: java/flight/flight-jdbc-driver/src/test/java/org/apache/arrow/driver/jdbc/FlightCookieTest.java [hunk 2]
```diff
@@ -0,0 +1,130 @@
...
+  @Test
+  public void testeCookies() throws SQLException {
+    try (Connection connection = FLIGHT_SERVER_TEST_RULE.getConnection()) {
+
+      // Since the first cookie was just updated by the client it was not send yet.
+      // That's why we're checking for null.
+      Assert.assertNull(FLIGHT_SERVER_TEST_RULE.getFactory().getCookie());
+
+      // Run another action for check if the cookies was sent by the client.
+      Statement statement = connection.createStatement();
```
Comment (jduo) [inline java/flight/flight-jdbc-driver/src/test/java/org/apache/arrow/driver/jdbc/FlightCookieTest.java:114]: try with resources for both the Statement and the Catalogs ResultSet. If the test fails, the close() calls won't happen.

File: java/flight/flight-jdbc-driver/src/test/java/org/apache/arrow/driver/jdbc/FlightServerTestRule.java [hunk 3]
```diff
@@ -231,4 +249,55 @@ public FlightServerTestRule build() {
...
+    public void onCallErrored(Throwable throwable) {
+
+    }
+
+    /**
+     * A factory for the MiddlewareCookkie.
+     */
+    static class Factory implements FlightServerMiddleware.Factory<MiddlewareCookie> {
+
+      private boolean receivedCookieHeader = false;
```
Comment (jduo) [inline java/flight/flight-jdbc-driver/src/test/java/org/apache/arrow/driver/jdbc/FlightServerTestRule.java:287]: This doesn't seem to get used.

File: java/flight/flight-jdbc-driver/src/test/java/org/apache/arrow/driver/jdbc/FlightCookieTest.java [hunk 4]
```diff
@@ -0,0 +1,130 @@
...
+        listener.error(throwable);
+      } finally {
+        listener.completed();
+      }
+    };
+    FLIGHT_SQL_PRODUCER.addCatalogQuery(commandGetTableTypes, commandGetTableTypesResultProducer);
+  }
+
+  @Test
+  public void testeCookies() throws SQLException {
```
Comment (vfraga) [inline java/flight/flight-jdbc-driver/src/test/java/org/apache/arrow/driver/jdbc/FlightCookieTest.java:106]:
```suggestion
  public void testCookies() throws SQLException {
```

File: java/flight/flight-jdbc-driver/src/main/java/org/apache/arrow/driver/jdbc/client/ArrowFlightSqlClientHandler.java (hunk 1 above)
Comment (iurysalino) [reply java/flight/flight-jdbc-driver/src/main/java/org/apache/arrow/driver/jdbc/client/ArrowFlightSqlClientHandler.java:519]: fixed

File: java/flight/flight-jdbc-driver/src/test/java/org/apache/arrow/driver/jdbc/FlightCookieTest.java (hunk 4 above)
Comment (iurysalino) [reply java/flight/flight-jdbc-driver/src/test/java/org/apache/arrow/driver/jdbc/FlightCookieTest.java:106]: fixed

File: java/flight/flight-jdbc-driver/src/test/java/org/apache/arrow/driver/jdbc/FlightCookieTest.java (hunk 2 above)
Comment (iurysalino) [reply java/flight/flight-jdbc-driver/src/test/java/org/apache/arrow/driver/jdbc/FlightCookieTest.java:114]: fixed

File: java/flight/flight-jdbc-driver/src/test/java/org/apache/arrow/driver/jdbc/FlightServerTestRule.java (hunk 3 above)
Comment (iurysalino) [reply java/flight/flight-jdbc-driver/src/test/java/org/apache/arrow/driver/jdbc/FlightServerTestRule.java:287]: The `receivedCookieHeader` are used in `onBeforeSendingHeaders`. Line 266.

**Pull request merged**